### PR TITLE
Add electronic horizon example that shows upcoming intersections

### PR DESCRIPTION
**Pull request by azarovalex**
Fixes: #88.
Video: 

https://user-images.githubusercontent.com/1976216/114015981-09529480-9873-11eb-85c5-af46f272b8e2.mp4

Explanations:
This example draws the current most probable path and shows the name of the cross street if there is an upcoming intersection.
Originally we wanted to copy-paste code from the Example app that we use for development, however, [the code there](https://github.com/mapbox/mapbox-navigation-ios/blob/release-v2.0/Example/ViewController%2BFreeDrive.swift#L100) prints the name of the next outlet edge, which is usually the same as the name of the current edge. And it also lacks the visualization of the route, which I added in this example.

### CODE REVIEW

File: Navigation-Examples/Examples/Upcoming-Intersection.swift [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        updateLabel(currentStreetName: currentStreetName, predictedCrossStreet: upcomingCrossStreet)
+
+        // Drawing the most probable path
+        let mostProbablePath = routeLine(from: horizonTree.start, roadGraph: passiveLocationManager.dataSource.roadGraph)
+        updateMostProbablePath(with: mostProbablePath)
+    }
+
+    private func streetName(for edge: ElectronicHorizon.Edge) -> String? {
+        let edgeMetadata = passiveLocationManager.dataSource.roadGraph.edgeMetadata(edgeIdentifier: edge.identifier)
+        return edgeMetadata?.names.first.flatMap { roadName in
```
Comment (bamx23) [inline Navigation-Examples/Examples/Upcoming-Intersection.swift:83]: Why `flatMap`?

File: Navigation-Examples/Examples/Upcoming-Intersection.swift [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        }
+    }
+
+    private func nearestCrossStreetName(from edge: ElectronicHorizon.Edge) -> String? {
+        let initialStreetName = streetName(for: edge)
+        var currentEdge: ElectronicHorizon.Edge? = edge
+        while let nextEdge = currentEdge?.outletEdges.max(by: { $0.probability < $1.probability }) {
+            let nextStreetName = streetName(for: nextEdge)
+            if nextStreetName != nil && nextStreetName != initialStreetName {
```
Comment (bamx23) [inline Navigation-Examples/Examples/Upcoming-Intersection.swift:98]:
```suggestion
            if let nextStreetName = streetName(for: nextEdge), nextStreetName != initialStreetName {
```

File: Navigation-Examples/Examples/Upcoming-Intersection.swift [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            if let shape = roadGraph.edgeShape(edgeIdentifier: currentEdge.identifier) {
+                coordinates.append(contentsOf: shape.coordinates.dropFirst(coordinates.isEmpty ? 0 : 1))
+            }
+            edge = currentEdge.outletEdges.max(by: { $0.probability < $1.probability })
+        }
+        return coordinates
+    }
+
+    private func updateMostProbablePath(with mostProbablePath: [CLLocationCoordinate2D]) {
+        let feature = Feature(geometry: .multiLineString(MultiLineString([mostProbablePath])))
```
Comment (bamx23) [inline Navigation-Examples/Examples/Upcoming-Intersection.swift:140]: You don't need multiline here. `.lineString` is enough

File: Navigation-Examples/Examples/Upcoming-Intersection.swift [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    func setupNavigationMapView() {
+        navigationMapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+        navigationMapView.mapView.locationManager.overrideLocationProvider(with: passiveLocationManager)
+        navigationMapView.mapView.update {
+            $0.location.puckType = .puck2D()
+        }
+        navigationMapView.mapView.on(.styleLoaded) { _ in
+            self.setupMostProbablePathStyle()
```
Comment (bamx23) [inline Navigation-Examples/Examples/Upcoming-Intersection.swift:29]: It's better to use `[weak self]` in this closure.

File: Navigation-Examples/Examples/Upcoming-Intersection.swift (hunk 1 above)
Comment (azarovalex) [reply Navigation-Examples/Examples/Upcoming-Intersection.swift:83]: I can use `if let` but there will be two return statements

**Review by bamx23 — APPROVED** [review]

**Pull request merged**